[QoS] Implement TODO to switch endpoint store to a field on service state

commoddity opened the pull request:
## 🌿 Summary

Refactor EVM QoS implementation to move endpoint store functionality into service state and consolidate related logic.

Implements this TODO from PR #200:
```go
// TODO_TECHDEBT(@commoddity, @adshmh): endpoint store should live in the serviceState struct and not the other way around.
```

### 🌱 Primary Changes:
- Move endpoint store functionality into `serviceState` struct as a field
- Consolidate endpoint selection and validation logic within `serviceState`
- Remove standalone `endpointStore` implementation and related files

### 🍃 Secondary changes:
- Update request context to use `serviceState` instead of `endpointStore`

## 🛠️ Type of change

Select one or more from the following:

- [ ] Code health or cleanup

## 🤯 Sanity Checklist

- [ ] I have updated the GitHub Issue 'assignees', 'reviewers', 'labels', 'project', 'iteration' and 'milestone'
- [ ] For code, I have run 'make test_all'

adshmh commented on qos/evm/qos.go:
```diff
@@ -33,22 +61,54 @@ func NewQoSInstance(logger polylog.Logger, config EVMServiceQoSConfig) *QoS {
+		return errors.New("ApplyObservations: received nil EVM observation")
+	}
+
+	updatedEndpoints := q.endpointStore.updateEndpointsFromObservations(
```
Please revisit this section: if endpointStore is part of the service state (a good change), these details need to be handled in the serviceState: i.e. observations should be passed to a single serviceState method, which handles both endpoint updates and serviceState updates.

commoddity replied:
Makes sense; I believe this can be remedied by just moving the method to the service state struct (which still allows `QoS` to implement `gateway.QoSService` as `serviceState` is embedded). I've made this change.

adshmh commented on qos/evm/service_state.go:
```diff
@@ -0,0 +1,278 @@
+// It has three main responsibilities:
+//  1. Generate QoS endpoint checks for the hydrator
+//  2. Select a valid endpoint for a service request
+//  3. Update the service state from endpoint observations
```
3. Update the stored endpoints from observations.
4. Update the stored service state from observations.

adshmh commented on qos/evm/service_state.go:
```diff
@@ -0,0 +1,278 @@
+}
+
+// shouldChainIDCheckRun returns true if the chain ID check is not yet initialized or has expired.
+func (ss *serviceState) shouldChainIDCheckRun(check endpointCheckChainID) bool {
```
Nice!

adshmh commented on qos/evm/service_state.go:
```diff
@@ -0,0 +1,278 @@
+
+// serviceState provides the endpoint selection capability required
+// by the protocol package for handling a service request.
+var _ protocol.EndpointSelector = &serviceState{}
```
I suggest moving this section to an `endpoint_selection.go` file to make it easy to find the logic.
This would make it easier for someone with little context to a) find the endpoint selection logic and b) recognize how it is related to the service state.

adshmh commented on qos/evm/service_state.go:
```diff
@@ -0,0 +1,278 @@
+// - The endpoint's response to an `eth_chainId` request is not the expected chain ID.
+// - The endpoint's response to an `eth_blockNumber` request is greater than the perceived block number.
+// - The endpoint's archival check is invalid, if enabled.
+func (ss *serviceState) validateEndpoint(endpoint endpoint) error {
```
Great refactor: consolidating endpoint validation logic in serviceState. This will help make `endpoint` merely a data container.

adshmh reviewed (changes requested):
Great work! Just a few comments to address, mostly re organizing the selection/update logic.

commoddity commented:
> Great work! Just a few comments to address, mostly re organizing the selection/update logic.

Thanks for the review @adshmh ; I believe I have addressed the comments.

adshmh reviewed (approved):
Great PR!
